Pluggable message routing

knolleary opened the pull request:
Starting design for the pluggable messaging routing. This is a work in progress, so the document contains placeholders and open questions that will be resolved prior to final review/merge.

knolleary commented:
With the recent replanning, this item is no longer in-plan for 1.0. This design will remain as a work in progress for the time being.

knolleary commented:
I've marked this as Ready For Review, although it is still being worked on. But as this is now actively being worked on, this PR should reflect that.

dceejay commented on designs/pluggable-message-routing/README.md:
```diff
@@ -0,0 +1,263 @@
+    "source": {
+        "id": "<node-id>",
+        "node": "<node-object>",
+        "port": "<index of port being sent on>",
```
what does port mean in this instance ?

knolleary replied:
It identifies which output port (`0`-indexed) the message was sent.

dceejay replied:
So if on port 1 or above it is already cloned ?

knolleary replied:
Any built-in cloning that happens will occur between the `preRoute` and `onSend` stages - and will only occur if the `cloneMessage` property is `true`. The port number has no direct bearing on the logic here.

For example, if a node sends a single message out of port 2 and there's only one wire connected to that port, then the node will set `cloneMessage` to `false` because it needs no further cloning.

dceejay commented on designs/pluggable-message-routing/README.md:
```diff
@@ -0,0 +1,263 @@
+
+
+##### Scenario: remote message routing
+
```
in this instance - does destination just have more properties such that it knows what to do - or is the `node` property replaced by a "remote" target definition ?

knolleary replied:
At the `preRoute` step, we only have the `id` of the node it has to send to. It is up to the handler to decide what it means to send to that `id`. Other runtime APIs would need to be provided to let it 'see' the full flow configuration and determine if any given `id` is local or remote and if there is any additional meta data embedded in the flow (such as group properties). Those other APIs still need to be designed.

dceejay replied:
Ok

dceejay commented on designs/pluggable-message-routing/README.md:
```diff
@@ -0,0 +1,263 @@
+ - `onReceive` - verify message passed to a node
+ - `preSend` - verify the message(s) sent by a node
+ - `preRoute` - pass the message to a Test Stub node instead of the real node
+
```
while this is for testing - could it also answer the itch re runtime type checking  if suitable schemas could be provided "everywhere" to  check against ?

knolleary replied:
Anyone could write a hook that does anything. So yes, a dedicated message validation hook could well be created.

dceejay commented on designs/pluggable-message-routing/README.md:
```diff
@@ -0,0 +1,263 @@
+6. `postReceive` - the message has been passed to the node's `input` handler
+
+7. `onDone`, `onError` - the node has completed with a message or logged an error
+
```
Do there need to be anything re onDeploy and onClose also for the hook handlers to get hold of ? eg to initialise/close a remote connection ?

knolleary replied:
I haven't got to that part of the lifecycle yet :) But yes, this is opening up a broader hook-based model in the runtime.

With the old Router model, yes the router layers would get notified whenever the flow configuration changed. But it then left the layer to do all its own parsing and figuring out what's what in the flow. And many layers wouldn't care - an enhanced logger for example wouldn't need to do anything, it just wants the message events. I think this is where a more standard Flow API would be helpful (which may pull in the flow manipulation API piece from the flow linter work).

But yes, this is one of the next areas to look at in the design.

knolleary commented:
I've updated the hook names to make them a bit more consistent.

![](https://raw.githubusercontent.com/node-red/designs/640fda78c4ef059b1470a9ea4f73225ad0e11bb5/designs/pluggable-message-routing/message-router-events.png)

Added info on the runtime APIs and some words on the differences between `RED.events` and `RED.hooks`.

knolleary commented:
The implementation for this has now been merged in for the 1.2 release.

Going to merge this version of the design so that subsequent updates (such as exposing hooks via the settings.js file) can be managed as separate PRs.
